recipes/llm-voice-assistant/ios/PicoLLMVoiceAssistantDemo/ViewModel.swift

```diff
@@ -0,0 +1,335 @@
+
+                setStatusText("Loading picoLLM...")
+                picollm = try PicoLLM(accessKey: ACCESS_KEY, modelPath: selectedModelUrl!.path)
+                dialog = try picollm!.getDialog()
+
+                setStatusText("Loading Orca...")
+                let orcaModelPath = Bundle(for: type(of: self)).path(forResource: "orca_params_female", ofType: "pv")!
+                orca = try Orca(accessKey: ACCESS_KEY, modelPath: orcaModelPath)
+
+                setStatusText("Loading Audio Player...")
+                audioStream = try AudioPlayerStream(sampleRate: Double(self.orca!.sampleRate!))
+
+                setStatusText("Loading Voice Processor...")
+                VoiceProcessor.instance.addFrameListener(VoiceProcessorFrameListener(audioCallback))
+                VoiceProcessor.instance.addErrorListener(VoiceProcessorErrorListener(errorCallback))
+                try VoiceProcessor.instance.start(
+                    frameLength: Porcupine.frameLength,
+                    sampleRate: Porcupine.sampleRate)
+
+                DispatchQueue.main.async { [self] in
+                    enginesLoaded = true
+                }
+            } catch {
+                DispatchQueue.main.async { [self] in
+                    unloadEngines()
+                    errorMessage = "\(error.localizedDescription)"
+                }
+            }
+
+            DispatchQueue.main.async { [self] in
+                selectedModelUrl!.stopAccessingSecurityScopedResource()
+
+                modelLoadStatusText = ViewModel.modelLoadStatusTextDefault
+                enableLoadModelButton = true
+            }
+        }
+    }
+
+    public func unloadEngines() {
+        do {
+            try VoiceProcessor.instance.stop()
+            VoiceProcessor.instance.clearFrameListeners()
+            VoiceProcessor.instance.clearErrorListeners()
+        } catch {
+            DispatchQueue.main.async { [self] in
+                errorMessage = "\(error.localizedDescription)"
+            }
+        }
+
+        if porcupine != nil {
+            porcupine!.delete()
+        }
+        if cheetah != nil {
+            cheetah!.delete()
+        }
+        if picollm != nil {
+            picollm!.delete()
+        }
+        if orca != nil {
+            orca!.delete()
+        }
+        porcupine = nil
+        cheetah = nil
+        picollm = nil
+        orca = nil
+
+        errorMessage = ""
+        promptText = ""
+        chatText.removeAll()
+        enginesLoaded = false
+    }
+
+    private var completionQueue = DispatchQueue(label: "text-stream-queue")
+    private var completionArray: [String] = []
+
+    private func streamCallback(completion: String) {
+        DispatchQueue.main.async { [self] in
+            completionQueue.async {
+                self.completionArray.append(completion)
+            }
+            chatText[chatText.count - 1].append(text: completion)
```

Comment:
since we have the list of stop phrases, I filter them out so that we don't print or speak them